#----------------Dir_settings:----------------------------------------------
base_dir = '/mnt/share/'#where it start in docker env
main_dir = base_dir + 'repo/example_dataset/'#docs / repository in docker env
genome_dir = 'genome_b37/'# genome data
sample_dir = 'single_sample/'#sample directory
output_dir = 'output/'#output folder
logs_dir = 'logs/'# for logs...
jar_dir = '/opt/conda/opt/'#needed for the gatk3_register
#------------------------------------------------------------------------
#configfile: "config.yaml"
GENOME = "b37_human_g1k_v37_decoy.chr20"
KS1 = "b37_dbsnp_138.b37.chr_20p12.1.vcf.gz"
KS2 = "b37_Mills_and_1000G_gold_standard.indels.b37.chr_20p12.1.vcf.gz"
KS3 = "b37_1000G_phase1.indels.b37.chr_20p12.1.vcf.gz"
#KS -> KnowSites
# ? make an input in cmd env ? 
#------------------------------------------------------------------------
rule bwa_index:
	input:
		fa = main_dir + genome_dir + GENOME + ".fasta"# GENOME = b37_human_g1k_v37_decoy.chr20
	output:
		main_dir + genome_dir + GENOME + '.fasta.sa'
	shell:
		'bwa index {input.fa}' #funciona

rule BWA_samblaster_sambamba:
	input:
	 	index = main_dir + genome_dir + GENOME + '.fasta.sa',# GENOME = b37_human_g1k_v37_decoy.chr20
		ref = main_dir + genome_dir + GENOME + '.fasta',
		r_fow = main_dir + sample_dir + '{sample}.R1.fastq.gz',
		r_rev = main_dir + sample_dir + '{sample}.R2.fastq.gz'
	output:
		align = main_dir + '{sample}.alignment.bam'
		#log = 'mnt/share/repo/example_dataset/BSS.NA12878.20p12.1.30x.log'
	threads: 4
	shell:
		'bwa mem -t {threads} -R "@RG\\tID:NA12878\\tPL:Illumina\\tLB:NA12878\\tSM:NA12878\\tCN:Lab" -M {input.ref} {input.r_fow} {input.r_rev} | '
		'samblaster -M |'
		'sambamba view -S -f bam -l 0 /dev/stdin |'#/dev/stdin 
		'sambamba sort -m 4G -t {threads} -o {output.align} /dev/stdin 2>&1 ' #> {output.log}' #/dev/stdin 


rule gatk3_register:
	input:
		base_dir + 'GenomeAnalysisTK-3.8-1-0-gf15c1c3ef/GenomeAnalysisTK.jar'
		#base_dir = '/home/jmatos/docker_pipe/'
	output:
		jar_dir +'gatk-3.8/GenomeAnalysisTK.jar'
		#jar_dir = '/opt/conda/opt/'
	shell:
		'gatk-register {input}'

rule picard_dict:
	input:
		main_dir + genome_dir + GENOME + '.fasta'
	output:
		main_dir + genome_dir + GENOME + '.dict'
	shell:
		'picard CreateSequenceDictionary R={input} O={output}'

rule samtools_index:
	input:
		main_dir + genome_dir + GENOME + '.fasta'
	output:
		main_dir + genome_dir + GENOME + '.fasta.fai'
	shell:
		'samtools faidx {input}'

rule gatk_BaseRecalibrator:
	input:
		jar = jar_dir + 'gatk-3.8/GenomeAnalysisTK.jar',
		ref = main_dir + genome_dir + GENOME + '.fasta',
		fai = main_dir + genome_dir + GENOME + '.fasta.fai',
		dict = main_dir + genome_dir + GENOME + '.dict',
		ksites1 = main_dir + genome_dir + KS1,
		ksites2 = main_dir + genome_dir + KS2,#
		ksites3 = main_dir + genome_dir + KS3,
		#reads = main_dir + 'single_sample/NA12878.20p12.1.30x.bam'#single_sample
		reads = main_dir + '{sample}.alignment.bam' #single_sample
		#L = intervalfile
		#cap = '~/hg19/"$capture'#
	output:
		recal_data = main_dir + output_dir + '{sample}.sorted.bam_pre_recal_data.grp',#"$sample"
		base_recal = main_dir + output_dir + logs_dir +'{sample}.base_recal.qsub'#/logs

	shell:
		'java -Xmx4g -jar {input.jar} \
		-T BaseRecalibrator -R {input.ref} -nct 2 \
		-knownSites {input.ksites1} \
		-knownSites {input.ksites2} \
		-knownSites {input.ksites3} \
		-I {input.reads} \
		-o {output.recal_data} --interval_padding 100 \
		> {output.base_recal}'
		# -L {input.cap}  \

rule gatk_BaseRecalibrator2:#A
	input:
		jar = jar_dir +'gatk-3.8/GenomeAnalysisTK.jar',
		ref = main_dir + genome_dir + GENOME + '.fasta',
		fai = main_dir + genome_dir + GENOME + '.fasta.fai',
		dict = main_dir + genome_dir + GENOME + '.dict',
		ksites1 = main_dir + genome_dir + KS1,
		ksites2 = main_dir + genome_dir + KS2,#
		ksites3 = main_dir + genome_dir + KS3,
		bqsr = main_dir + output_dir +'{sample}.sorted.bam_pre_recal_data.grp',#"$sample"		
		#reads = main_dir + 'single_sample/NA12878.20p12.1.30x.bam'#single_sample
		reads = main_dir + '{sample}.alignment.bam'#single_sample
		#cap = '~/hg19/"$capture'#
	output:
		pos_recal_data = main_dir + output_dir +'{sample}.sorted.bam_pos_recal_data.grp',#"$sample"
		base_recal = main_dir + output_dir + logs_dir +'{sample}.base_recal_2.qsub'#/logs
	shell:
		'java -Xmx4g -jar {input.jar} \
		-T BaseRecalibrator -R {input.ref} -nct 2 \
		-knownSites {input.ksites1} \
		-knownSites {input.ksites2} \
		-knownSites {input.ksites3} \
		-BQSR {input.bqsr} \
		-I {input.reads} \
		-o {output.pos_recal_data} --interval_padding 100 \
		> {output.base_recal}'
		# -L {input.cap}  \

rule Analyze_Covariates:#A
	input:
		jar = jar_dir +'gatk-3.8/GenomeAnalysisTK.jar',
		fa = main_dir + genome_dir + GENOME + ".fasta",
		bqsr = main_dir + output_dir +'{sample}.sorted.bam_pre_recal_data.grp',
		pos_recal_data = main_dir + output_dir +'{sample}.sorted.bam_pos_recal_data.grp'	
	output:
		pdf = main_dir + output_dir + logs_dir + '{sample}.gen.sorted.bam_plots.pdf'
	shell:
		'java -Xmx2g -jar {input.jar} \
		-T AnalyzeCovariates \
		-R {input.fa} \
		-before {input.bqsr} \
		-after {input.pos_recal_data} \
		-plots {output.pdf}'

rule Apply_recalibration:
	input:
		jar = jar_dir +'gatk-3.8/GenomeAnalysisTK.jar',
		fa = main_dir + genome_dir + GENOME + ".fasta",
		bqsr = main_dir + output_dir +'{sample}.sorted.bam_pre_recal_data.grp',
		reads = main_dir + '{sample}.alignment.bam'#,
		#pdf = main_dir + output_dir + logs_dir + 'gen.sorted.bam_plots.pdf'
	output:
		recal = main_dir + output_dir + '{sample}.sorted.nodups.realigned.recal.bam',
		apply_recal = main_dir + output_dir + logs_dir + '{sample}.apply_recal.qsub'
	shell:
		'java -Xmx4g -jar {input.jar} \
		-T PrintReads \
		-R {input.fa} --filter_bases_not_stored -nct 8 \
		-I {input.reads} \
		-BQSR {input.bqsr} \
		-o {output.recal} \
		> {output.apply_recal}'

rule Call_Variants:
	input:
		jar = jar_dir +'gatk-3.8/GenomeAnalysisTK.jar',
		fa = main_dir + genome_dir + GENOME + ".fasta",
		#target = main_dir + genome_dir +'capture',# in case of exÃµes (WES vs WGS) #PROBLEM???
		recal = main_dir + output_dir + '{sample}.sorted.nodups.realigned.recal.bam',
		ksites1 = main_dir + genome_dir + KS1
	output:
		site = main_dir + output_dir + '{sample}.gvcf.gz',
		logs = main_dir + output_dir + logs_dir + '{sample}.call_vars.qsub'
	shell:
		'java -Xmx4g -jar {input.jar} \
		-T HaplotypeCaller -rf BadCigar -R {input.fa} \
		-ip 100 \
		-I {input.recal} \
		-stand_call_conf 10  --filter_bases_not_stored --emitRefConfidence GVCF \
		-D {input.ksites1} -variant_index_type LINEAR -variant_index_parameter 128000 \
		-o {output.site} > {output.logs}'